_posts/2019-12-20-differential-privacy/differential_privacy.Rmd

```diff
@@ -0,0 +1,470 @@
+
+2. Before updating the parameters, noise is added to the gradients, thus implementing the main idea of $\epsilon$-DP algorithms.
+
+In addition to $\epsilon$-DP optimization, TF Privacy provides _privacy accounting_. We'll see all this applied after an introduction to our example dataset.
+
+
+
+## Dataset
+
+The dataset we'll be working with[@ReissISL19], downloadable from the [UCI Machine Learning Repository](https://archive.ics.uci.edu/ml/datasets/PPG-DaLiA), is dedicated to heart rate estimation via [photoplethysmography](https://en.wikipedia.org/wiki/Photoplethysmogram).
+Photoplethysmography (PPG) is an optical method of measuring blood volume changes in the microvascular bed of tissue, which are indicative of cardiovascular activity. More precisely,
+
+> The PPG waveform comprises a pulsatile ('AC') physiological waveform attributed to cardiac synchronous changes in the blood volume with each heart beat, and is superimposed on a slowly varying ('DC') baseline with various lower frequency components attributed to respiration, sympathetic nervous system activity and thermoregulation. [@Allen_2007]
+
+In this dataset, heart rate determined from EKG provides the ground truth; predictors were obtained from two commercial devices, comprising PPG, electrodermal activity, body temperature as well as accelerometer data. Additionally, a wealth of contextual data is available, ranging from age, height, and weight to fitness level and type of activity performed.
+
+With this data, it's easy to imagine a bunch of interesting data-analysis questions; however here our focus is on differential privacy, so we'll keep the setup simple. We will try to predict heart rate given the physiological measurements from one of the two devices, Empatica E4. Also, we'll zoom in on a single subject, _S1_, who will provide us with 4603 instances of two-second heart rate values. ^[Relative files per subject are 1.4G in size.]
+
+As usual, we start with the required libraries; unusually though, as of this writing we need to disable version 2 behavior in TensorFlow, as TensorFlow Privacy does not yet fully work with TF 2. (Hopefully, for many future readers, this won't be the case anymore.)
+Note how TF Privacy -- a Python library -- is imported via `reticulate`.
+
+```{r}
+library(tensorflow)
+tf$compat$v1$disable_v2_behavior()
+
+library(keras)
+library(tfdatasets)
+library(tfautograph)
+
+library(purrr)
+
+library(reticulate)
+# if you haven't yet, install TF Privacy, e.g. using reticulate:
+# py_install("tensorflow_privacy")
+priv <- import("tensorflow_privacy")
+```
+
+From the downloaded archive, we just need `S1.pkl`, saved in a [native Python serialization format](https://docs.python.org/3.8/library/pickle.html), yet nicely loadable using `reticulate`:
```

Comment:
Understanding that the focus of this blog post is DP and not data ingest... is there a CSV version of the data? If it's not too much work to start from a CSV I think you'd be able to relate to a lot more R users.

Reply:
I see, but on the other hand:

- It also makes sense to show we can handle inconvenient file formats found "in the wild"
- we want to show tfdatasets pipelines as _the_ way to do preprocessing for DL, if possible
- especially for that topic of transforming to `timesteps`, I think that once one has understood how to do this with `dataset_window` it's actually the easiest way (ok, rephrasing: once one has an example to copy-paste :-)) - myself, for example, I'll definitely stick to this way now that I've found out ;-))

BUT, I do think you might be writing this having some other project in mind. There, it will definitely be a thing to consider, and we should discuss the target group thing in more detail next month.